R/commonequivalencefunctions.R

```diff
@@ -505,8 +505,8 @@
   # BF_equivalence = density in the range of the posterior / density in the range of the prior
 
   # Step 1: Density in the range of the prior
-  prior <- metaBMA::prior("norm", c(mean = prior.mean, sd = sqrt(prior.variance)))
-  integralEquivalencePrior <- integrate(prior, lower = options$lowerbound, upper = options$upperbound)
+  #prior <- metaBMA::prior("norm", c(mean = prior.mean, sd = sqrt(prior.variance)))
+  integralEquivalencePrior <- integrate(dnorm, mean = prior.mean, sd = sqrt(prior.variance), lower = options$lowerbound, upper = options$upperbound)
```

Comment:
I'd use `pnorm` over `integrate(dnorm, ...)`. That should be faster and more accurate.
``` r
mean  <-  3.123
sd    <-  2.123
lower <- -1.123
upper <-  3.123

integrate(dnorm, mean = mean, sd = sd, lower = lower, upper = upper)
#> 0.4772499 with absolute error < 5.3e-15
pnorm(upper, mean, sd) - pnorm(lower, mean, sd)
#> [1] 0.4772499

# speed
bch <- bench::mark(
  expr1 = integrate(dnorm, mean = mean, sd = sd, lower = lower, upper = upper)$value,
  expr2 = pnorm(upper, mean, sd) - pnorm(lower, mean, sd)
)
bch
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 expr1        26.1µs   32.1µs    29526.    4.05KB     38.4
#> 2 expr2         2.4µs    2.9µs   272881.        0B      0
summary(bch, relative = TRUE)
#> # A tibble: 2 x 6
#>   expression   min median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <dbl>  <dbl>     <dbl>     <dbl>    <dbl>
#> 1 expr1       10.8   11.1      1          Inf      Inf
#> 2 expr2        1      1        9.24       NaN      NaN

# accuracy
mean  <-  0
sd    <-  1e-9
lower <- -10
upper <-  10

# very wrong answer
integrate(dnorm, mean = mean, sd = sd, lower = lower, upper = upper)
#> 0 with absolute error < 0
# correct answer
pnorm(upper, mean, sd) - pnorm(lower, mean, sd)
#> [1] 1
```

<sup>Created on 2021-07-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Reply:
That's way better, thank you! Somewhere else in the code I also use integrate, but then over a t distribution: 

```
prior <- metaBMA::prior("t", c(location = prior.location, scale = prior.scale, nu = prior.df))
integralEquivalencePrior <- integrate(prior, lower = options$lowerbound, upper = options$upperbound)
```

Is there also an alternative that works better for this?

Reply:
yeah you can do the same there:

``` r
lower    <- -3.456
upper    <-  2.12
location <-  2.12
scale    <-  0.65
df       <-  4

dt <- metaBMA::prior("t", c(location = location, scale = scale, nu = df))
integralEquivalencePrior <- integrate(dt, lower = lower, upper = upper)
integralEquivalencePrior
#> 0.4994929 with absolute error < 3e-05
LaplacesDemon::pst(upper, location, scale, df) - LaplacesDemon::pst(lower, location, scale, df)
#> [1] 0.4994929
```
<sup>Created on 2021-07-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

I'd use the package `LaplacesDemon` for this because `metaBMA` also uses that package to compute e.g., the density of the t-distribution.

Reply:
Btw you can do this for practically all univariate distributions, as long as the cumulative distribution function is defined.